packages/agents/cartographer/poller/src/adapters/database/client.ts

```diff
@@ -262,69 +252,75 @@ export const getTransfersWithDestinationPending = async (
-    `${orderDirection}`,
-  )} LIMIT ${db.param(limit)}`.run(poolToUse);
+
+  const transfers = await db
+    .select(
+      "transfers",
+      db.sql<s.transfers.SQL>`
+      (${{ destination_domain: domain }} 
+      OR "destination_domain" IS NULL)
+      AND ("xcall_timestamp" IS NOT NULL AND ("execute_timestamp" IS NULL 
+      OR "reconcile_timestamp" IS NULL))
+      `,
+      { order: { by: "update_time", direction: orderDirection }, limit },
+    )
+    .run(poolToUse);
 
   const transfer_ids = transfers.map((transfer) => transfer.transfer_id);
   return transfer_ids;
 };
 
-export const saveRouterBalances = async (routerBalances: RouterBalance[], _pool?: Pool): Promise<void> => {
+export const saveRouterBalances = async (
+  routerBalances: RouterBalance[],
+  _pool?: Pool | db.TxnClientForRepeatableRead,
+): Promise<void> => {
   const poolToUse = _pool ?? pool;
   const routers: s.routers.Insertable[] = routerBalances.map((router) => {
     return { address: router.router };
   });
 
-  // TODO: make this a single query! we should be able to do this with postgres
-  for (const router of routers) {
-    await db.sql<s.routers.SQL, s.routers.JSONSelectable>`
-    INSERT INTO ${"routers"} (${db.cols(router)}) VALUES (${db.vals(
-      router,
-    )}) ON CONFLICT ("address") DO NOTHING RETURNING *
-    `.run(poolToUse);
-
-    const balances = (routerBalances.find((r) => r.router === router.address) ?? {}).assets ?? [];
-    const dbBalances: { balance: s.asset_balances.Insertable; asset: s.assets.Insertable }[] = balances.map((b) => {
-      return {
-        balance: {
-          asset_canonical_id: b.canonicalId,
-          asset_domain: b.domain,
-          router_address: router.address,
-          // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
-          balance: b.balance as any,
-        },
-        asset: {
-          local: b.local,
-          adopted: b.adoptedAsset,
-          canonical_id: b.canonicalId,
-          canonical_domain: b.canonicalDomain,
-          domain: b.domain,
-        },
-      };
-    });
-
-    for (const balance of dbBalances) {
-      await db.sql<s.assets.SQL, s.assets.JSONSelectable>`
-      INSERT INTO ${"assets"} (${db.cols(balance.asset)}) VALUES (${db.vals(
-        balance.asset,
-      )}) ON CONFLICT ("canonical_id", "domain") DO UPDATE SET (${db.cols(balance.asset)}) = (${db.vals(
-        balance.asset,
-      )}) RETURNING *
-    `.run(poolToUse);
-
-      await db.sql<s.asset_balances.SQL, s.asset_balances.JSONSelectable>`
-      INSERT INTO ${"asset_balances"} (${db.cols(balance.balance)}) VALUES (${db.vals(
-        balance.balance,
-      )}) ON CONFLICT ("asset_canonical_id", "asset_domain", "router_address") DO UPDATE SET (${db.cols(
-        balance.balance,
-      )}) = (${db.vals(balance.balance)}) RETURNING *
-    `.run(poolToUse);
+  await db.repeatableRead(poolToUse, async (txnClient) => {
```

Comment:
It looks like this function is called from `routers.ts`: https://github.com/connext/nxtp/blob/96086f0270349aab37d9636d7b5f54e8ce3acf5b/packages/agents/cartographer/poller/src/lib/operations/routers.ts#L23

If a transactional scope is ever needed within `routers.ts`, it would be better to define the transaction there. This would allow to have a transactional scope that spans multiple db commands within `routers.ts`.

It would also keep this file cleaner so that the responsibility of defining transactions is moved to whoever is calling functions from `client.ts`.